### PR TITLE
Add a delay in process text changes to analyzer config and additional…

### PR DESCRIPTION
**Pull request by mavasani**
… documents

Fixes VSO [#1048866](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1048866): Editing .editorconfig file in large solution makes editor painfully slow, high CPU and RAM usage, sometimes crashes
Addresses #41837

Verified manually that typing in Roslyn.sln's .editorconfig is now fairly responsive. There is obviously a spike in CPU after the typing has finished and text changes are processed after the delay, but that is expected as the analyzer config options for entire solution has changed.

### CODE REVIEW

**Comment by mavasani** [comment]
@sharwell had another suggestion: only process changes to editorconfig on save. However, that breaks the scenario where a code fix updates .editorconfig (our code style/severity configuration code fixes), as the code fix application updates the .editorconfig but does not save the changes.

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs [hunk 1]
```diff
@@ -514,7 +514,8 @@ protected internal void OnAnalyzerConfigDocumentOpened(DocumentId documentId, So
 
                 var newSolution = this.SetCurrentSolution(currentSolution);
 
-                SignupForTextChanges(documentId, textContainer, isCurrentContext, onDocumentTextChanged);
+                // PERF: Use a 5-second delay to process text changes to additional files and analyzer config files.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs:517]: Would this cause the workspace to get out of sync with the current buffers then? I'd be pretty worried if that has downstream impacts.

**Comment by sharwell** [comment]
> but does not save the changes.

It does save the changes if the file is not currently open

**Comment by jasonmalinowski** [comment]
@mavasani What type of edit was happening? Can we instead adjust the code here to simply not re-create syntax trees in this case?

https://github.com/dotnet/roslyn/blob/6c1584c0039c1a1cdfe337eddac21197cbea4f04/src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs#L432-L433

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs (hunk 1 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs:517]: This is absolutely a question where the answer also has to come from you :-) I have no idea, but trust your call - if you feel this is risky then we can take a different approach.

**Comment by mavasani** [comment]
> What type of edit was happening?

This is just normal typing inside a repo level .editorconfig.

>  Can we instead adjust the code here to simply not re-create syntax trees in this case?

I will investigate, thanks.

File: src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs:517]: My first thought was that this was terrifyingly risky. But we already dealt with races like this in the regular source files with our GetDocumentInCurrentContextWithChanges...and we don't have an equivalent for anything else. So I suspect either:

1. Nothing was assuming this today, so this is fine.
2. Something is assuming this today but was possibly already busted.

And if it's 2, the fix would be to perhaps introduce a WithChanges to fix it either way, but that might undo this optimization then.

**Comment by jasonmalinowski** [comment]
@mavasani if we touch that other code and at least not reparse if the .editorconfig rules didn't change, that'd probably save a lot. But there might still be some expensive bits where you're typing in the name of the diagnostic and we still are generating new stuff all the time. So we might need both fixes...?

**Comment by CyrusNajmabadi** [comment]
@agocke

**Comment by mavasani** [comment]
> if we touch that other code and at least not reparse if the .editorconfig rules didn't change, that'd probably save a lot. But there might still be some expensive bits where you're typing in the name of the diagnostic and we still are generating new stuff all the time. So we might need both fixes...?

Yes, the scenario for manually typing a diagnostic severity entry and/or the option key/value pair, where either the trees or the analyzer config set will have to be regenerated with every keystroke made me take the route in this PR. I think we may need both fixes for optimal performance for this scenario. I'll keep #41837 active for the syntax tree optimization where we avoid re-parsing unless necessary. This change does seem to address the immediate user facing issue of extremely slow typing experience and subsequent OOMs. I will test this change a bit more to try answer your questions.

**Comment by agocke** [comment]
There was a previous proposal that we move the syntaxtree-specific options to actually be a provider on compilationoptions, analogous to what we do for AnalyzerConfigOptionsProvider. I think it's a fine solution and would probably address this problem. It's on my backlog, but unfortunately just hasn't risen in priority yet.

I'm open to workarounds until I can get that change through

**Comment by mavasani** [comment]
> There was a previous proposal that we move the syntaxtree-specific options to actually be a provider on compilationoptions, analogous to what we do for AnalyzerConfigOptionsProvider. I think it's a fine solution and would probably address this problem.

I believe that approach would resolve the performance issues for all typing scenarios inside editorconfig. @jasonmalinowski How about going with this delay based change as a temporary workaround, which gets reverted when @agocke makes the compiler change mentioned above?

**Comment by CyrusNajmabadi** [comment]
> I believe that approach would resolve the performance issues for all typing scenarios inside editorconfig. @jasonmalinowski How about going with this delay based change as a temporary workaround, 

Do we need the temporary workaround.  I agree things are not good now.  But that helps motivate doing things teh right way.  I'd put money on nothing getting done here in the future if we put in the workaround.

I personally don't like the workaround because i'm very nervous about inconsistencies in our models.  I feel like they will end up withhard to find and understand bugs.

**Comment by mavasani** [comment]
>  But that helps motivate doing things teh right way. I'd put money on nothing getting done here in the future if we put in the workaround.

Personally, I don't think the customer cares, and given this issue is a big reliability issue that causes VS to crash/OOM, it is more important for us to address the customer issue then whether or not this gets done in the right way. I understand your stance, but IMO that is an acceptable stance only for non-reliability and non-performance critical issue. I would be fine with either of the following:

1. Prioritize the compiler work for 16.7
2. Go with the current proposed workaround in IDE layer.
3. Propose an alternate approach/workaround in the IDE layer that addresses these performance issues.

NOTE: The VS feedback issue has been escalated as a high pri reliability issue, impacting more then one critical VS enterprise customers.

**Comment by jasonmalinowski** [comment]
> I'd put money on nothing getting done here in the future if we put in the workaround.

If we think the compiler fix is the right fix, we should just do it. "unfortunately just hasn't risen in priority yet" is now moot given we have high priority feedback ticket from customers. If @agocke doesn't personally have time then we can probably have somebody else pick up the work since I don't think there's a deep technical problem as much as just making sure we all agree on the appropriate shape. #19541 was the original PR that had the original shape and although it wouldn't be a great fit that was something we had previously designed.

**Comment by jasonmalinowski** [comment]
One other potential perf issue issue even with the delay approach is branch switches: we'll still be reparsing things multiple times if a git pull or branch switch changes the .editorconfig, and redoing the API shape would help with that as well.

File: src/Workspaces/Core/Portable/Workspace/Workspace.TextTracker.cs [hunk 2]
```diff
@@ -43,10 +56,51 @@ public void Disconnect()
...
+                {
+                    await Task.Delay(_delayBetweenProcessingTextChanges).ConfigureAwait(false);
+
+                    if (_lastTextChangeTickCount != currentTickCount)
+                    {
+                        // There was another subsequent text change.
+                        return;
+                    }
+
+                    OnTextChangedCore(e.NewText);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Workspace.TextTracker.cs:95]: Does this work also need to be cancelled if the Disconnect() method of this class is called?

**Comment by mavasani** [comment]
Thanks @jasonmalinowski. I think your last concern about perf concern on branch switches (which is also one of the core scenarios where we have received customer performance reports in Roslyn), is good enough to convince me that we should just prioritize and fix this in the compiler. Let me update the VSO ticket accordingly and close this PR.

**Comment by mavasani** [comment]
Closing this PR as it is determined that any workaround in IDE layer here is risky and can potentially introduce race conditions/hard to investigate bugs.

**Pull request closed without merging**